### PR TITLE
BUG: non-iterable value in meta raise error in json_normalize

### LINKED ISSUE #31507: json_normalize in 1.0.0 with meta path specified - expects iterable
#### Code Sample, a copy-pastable example if possible

```python
import json
from pandas.io.json import json_normalize

the_json = """                                                                  
[{"id": 99,                                                                     
  "data": [{"one": 1, "two": 2}]                                                
}]                                                                              
"""

print(json_normalize(json.loads(the_json),
                     record_path=['data'], meta=['id']))
```
#### Problem description

Through 0.25.3, this program generates a DataFrame with one row.  In 1.0.0 it fails with an exception:

```
Traceback (most recent call last):
  File "foo.py", line 11, in <module>
    record_path=['data'], meta=['id']))
  File "/home/dataczar/venvs/test/lib/python3.7/site-packages/pandas/util/_decorators.py", line 66, in wrapper
    return alternative(*args, **kwargs)
  File "/home/dataczar/venvs/test/lib/python3.7/site-packages/pandas/io/json/_normalize.py", line 327, in _json_normalize
    _recursive_extract(data, record_path, {}, level=0)
  File "/home/dataczar/venvs/test/lib/python3.7/site-packages/pandas/io/json/_normalize.py", line 314, in _recursive_extract
    meta_val = _pull_field(obj, val[level:])
  File "/home/dataczar/venvs/test/lib/python3.7/site-packages/pandas/io/json/_normalize.py", line 246, in _pull_field
    f"{js} has non iterable value {result} for path {spec}. "
TypeError: {'id': 99, 'data': [{'one': 1, 'two': 2}]} has non iterable value 99 for path ['id']. Must be iterable or null.
```

I don't see any documentation changes that suggest a backwards-incompatible change.  All my calls to `json_normalize` that don't use `meta` function as before.

#### Expected Output

Through 0.25.3, the output was:

```
   one  two  id
0    1    2  99
```

#### Output of ``pd.show_versions()``

From my virtualenv with pandas 1.0.0:

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.2.0-042stab120.16
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.0.0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.0.3
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : 1.3.13
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

From my virtualenv with 0.25.x:

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.2.0-042stab120.16
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 0.25.0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.0.3
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.7.0
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : 1.3.7
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #31507 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
Can you add a whatsnew for v1.0.1?

**Comment by charlesdong1991** [comment]
oops, i was editing whatsnew just now 😅

**Comment by charlesdong1991** [comment]
emm, 
```
self = <pandas.tests.io.json.test_normalize.TestNestedToRecord object at 0x14ab9d208>

    def test_meta_non_iterable(self):
        # GH 31507
        data = """[{"id": 99, "data": [{"one": 1, "two": 2}]}]"""
    
        result = json_normalize(json.loads(data), record_path=["data"], meta=["id"])
        expected_values = [[1, 2, "99"]]
        columns = ["one", "two", "id"]
        expected = DataFrame(expected_values, columns=columns)
>       tm.assert_frame_equal(result, expected)

pandas/tests/io/json/test_normalize.py:761: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/_libs/testing.pyx:65: in pandas._libs.testing.assert_almost_equal
    cpdef assert_almost_equal(a, b,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise_assert_detail(obj, msg, lobj, robj)
E   AssertionError: DataFrame.iloc[:, 2] (column name="id") are different
E   
E   DataFrame.iloc[:, 2] (column name="id") values are different (100.0 %)
E   [left]:  [99]
E   [right]: [99]
```

is there a bug in `assert_frame_equal`? 🤔  will check it out tomorrow

File: pandas/tests/io/json/test_normalize.py
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: The expected value here should be `99` not `"99"` (probably cause of CI error)

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/json/_normalize.py
Comment (jreback) [inline pandas/io/json/_normalize.py]: this is a code smell to add this. so it seems that result can be null, iterabe, or a scalar? if its a scalar what is the return value here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/json/_normalize.py
Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: you mean `[99]`? then will still return 99

i think this `Iterable` should only restrict if specifying `record_path`, but not for `meta`? Am I right about the behaviour here @WillAyd ?

File: pandas/tests/io/json/test_normalize.py
Comment (charlesdong1991) [reply pandas/tests/io/json/test_normalize.py]: moved! and wait for @WillAyd comment on the other one

File: pandas/io/json/_normalize.py
Comment (WillAyd) [reply pandas/io/json/_normalize.py]: I do agree that this is getting a little strange, especially since there is inspection of `meta` on line 273 of the same module and we are essentially repeating that here with a boolean indicator being manually supplied

@charlesdong1991 do you see a way to more logically order this function so we don't have to use this bool indicator?

Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: thanks for your reply! @WillAyd 

I will think about it a bit, i feel the patch for that `Iterable` in this `_pull_field` has code smell a bit because it is used for two cases which have different requirements

Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: > especially since there is inspection of meta on line 273 of the same module and we are essentially repeating that here with a boolean indicator being manually supplied

i just take a look again at the current codebase, seems that line 273 and onwards you referred is to validate/transform the key which is assigned to `meta` not the value the key associated with, in this case, `key` is `id` (and we could either specify it as `[id]` or `id` or more complex cases with nested list, and this part of code will deal with it), however, the code in `_pull_field` is to pull the value out, and in this case, the value is `99`, and for `record_path` the value should be an `Iterable`, while for `meta` it does not necessarily be the case, and therefore I think the patch added to check the type should only work for values of which the element of `record_path` point to, not for `meta`? @WillAyd

Comment (WillAyd) [reply pandas/io/json/_normalize.py]: @charlesdong1991 does splitting this into separate functions for record_path vs meta and just using this as a base for those functions make things cleaner?

**Comment by charlesdong1991** [comment]
@WillAyd @jreback I slightly changed the change made in #30145 and think this should be the cleanest solution, please take a look and any feedback is very appreciated!

**Comment by WillAyd** [comment]
I still find it pretty confusing; if we want to treat the meta differently than the record_path value why not just have a separate function or at least something to more clearly delineate? I think right now pretty confusing

**Comment by charlesdong1991** [comment]
```
if not isinstance(result, Iterable):
    if pd.isnull(result):
         result = []  # type: ignore
    else:
         raise TypeError(
               f"{js} has non iterable value {result} for path {spec}. "
               "Must be iterable or null.")
```
this patch was added in #30145 in `_pull_field` function, however, this will not only affect when specifying `record_path`, but also `meta`. And we would only like to make this patch for record_path, so my change here is to separate this patch out from `_pull_field` and only added it for record path. 

> if we want to treat the meta differently than the record_path value 

so actually we don't want to treat meta differently, but only treat record_path value differently

> why not just have a separate function or at least something to more clearly delineate?

I think this is basically what i did here in this PR? separate the added patch out, and assign it only for record path?

@WillAyd maybe i didn't fully understand your comment, please feel free to comment and let me know what you think, thanks a lot! 👍

**Comment by charlesdong1991** [comment]
@WillAyd any feedback comment? ^^

**Comment by WillAyd** [comment]
> this patch was added in #30145 in `_pull_field` function, however, this will not only affect when specifying `record_path`, but also `meta`

Right so I'm questioning that we even should do this. I think might be clearer if there were separate functions to handle each, which may dispatch to another for shared functionality.

**Comment by jbrockmendel** [comment]
Should this have a 1.0.2 label?

**Comment by charlesdong1991** [comment]
@WillAyd I made some change again following your suggestion, does this look slightly clearer?

**Comment by charlesdong1991** [comment]
@WillAyd would appreciate for some feedbacks! ^^

**Comment by laurentthebeginner** [comment]
@charlesdong1991, thanks for picking this one, i couldn't validate the transition to pandas 1.0.0 as it's impossible to parse data as it used to be.

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]:
```suggestion
    def _pull_field(js: Dict[str, Any], spec: Union[List, str]) -> Any:
```

Comment (WillAyd) [reply pandas/io/json/_normalize.py]: Does this only return a scalar?

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]:
```suggestion
    def _pull_records(js: Dict[str, Any], spec: Union[List, str]) -> Iterable:
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
I think this is heading in the right direction but still can clarify the intent a little more (some of this is a pre-existing problem, so thanks for being patient)

File: pandas/io/json/_normalize.py
Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: yeah, changed for `js` as well, since the result pulled out is the value of `js` which is currently defined with `Any`. thanks!

Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: emm, i changed to `Scalar`, but some annotation errors raised in `internal/manager.py`, and even if i changed back to `Any`, still see these errors:
```
pandas/core/internals/managers.py:1109: error: Unsupported target for indexed assignment
pandas/core/internals/managers.py:1111: error: Unsupported target for indexed assignment
pandas/core/internals/managers.py:1145: error: Unsupported target for indexed assignment
pandas/core/internals/managers.py:1146: error: Unsupported target for indexed assignment
pandas/core/internals/managers.py:1160: error: Unsupported target for indexed assignment
pandas/core/internals/managers.py:1161: error: Unsupported target for indexed assignment
```

do you have any idea how/why those errors are raised? any hints are very appreciated, thanks! @WillAyd

**Comment by WillAyd** [comment]
Those failures are unrelated so can ignore for now. There is another PR in flight to fix

**Comment by charlesdong1991** [comment]
many thanks for your quick response @WillAyd 

I will leave this PR then, and feel free to still take a look and see if anything else you want to see improvements. I will rebase once the fix PR is merged

**Comment by charlesdong1991** [comment]
any further feedbacks? @WillAyd 

I think this would be nice to be included in `1.0.2` as it seems impact many users based on the responds under the Issue

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]: Why did this need to change?

Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: emm, this is because i moved `type: ignore` to `result = result[field]` below, because without  `type: ignore`, this `result = result[field]` will raise an error in type annotation which complains the incompatible assignment. 

@WillAyd any thoughts on this?

**Review by TomAugspurger — APPROVED** [review]
LGTM, other than a question about the release note.

**Comment by charlesdong1991** [comment]
@TomAugspurger thanks for your quick feedback, however, maybe something went wrong with github, your question about release note is not presented, could you please rewrite your question? thanks!

**Comment by TomAugspurger** [comment]
Strange. Is this fixing a regression? If so, the note should be towards the top of 1.0.2.rst with the other regression fixes.

**Comment by charlesdong1991** [comment]
thanks! @TomAugspurger 
this is not a regression issue IMHO, and shall I keep it in `I/O` section?

**Comment by TomAugspurger** [comment]
Yep, that sounds good.

**Review by jreback — APPROVED** [review]
lgtm

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]: Can you advise specifically what the error is? I realize we want to get this in for 1.0.2 so not going to block, but I still think this code is suspect (not your fault - just a historical artifact) so I'd hate to suppress a warning about another bug that this fix could be introducing

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/json/_normalize.py
Comment (WillAyd) [reply pandas/io/json/_normalize.py]: Took a look at this locally; I think if you revert some of the other changes here you won't need the ignore. Suggested separately

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]:
```suggestion
    def _pull_field(js: Dict[str, Any], spec: Union[List, str]) -> Union[Scalar, Iterable]:
```

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]:
```suggestion
                result = result[field]
```

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]:
```suggestion
            result = result[spec]
```

File: pandas/io/json/_normalize.py
Comment (WillAyd) [inline pandas/io/json/_normalize.py]:
```suggestion
    def _pull_records(js: Dict[str, Any], spec: Union[List, str]) -> Iterable:
```

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/json/_normalize.py
Comment (charlesdong1991) [reply pandas/io/json/_normalize.py]: yeah, having `Any` instead of `Scalar` could fix the issue

**Comment by charlesdong1991** [comment]
Sorry about the mess I made on the type annotation and function naming  @WillAyd 

I have committed a fix following ur suggestion, pls let me know your thoughts, and I will try to update PR asap to get this done before release.

**Comment by TomAugspurger** [comment]
CI is passing, so I'm planning to merge this in an hour or so if there aren't any objections. We can fixup annotations later if needed.

**Review by WillAyd — APPROVED** [review]
lgtm. Appears the annotations were already cleaned up

**Comment by WillAyd** [comment]
Thanks @charlesdong1991 . Great job seeing this through

**Pull request merged**
